Kernel/Modules/AgentTicketCompose.pm

```diff
@@ -1264,60 +1375,316 @@ sub Run {
+                    $GetParam{StateID}               = '';
+                    $NewChangedElements{NextStateID} = 1;
+                    $ChangedStdFields{NextStateID}   = 1;
+                }
+
+                # autoselect
+                elsif ( !$GetParam{StateID} && $Autoselect && $Autoselect->{NextStateID} ) {
+                    $GetParam{StateID} = $FieldRestrictionsObject->Autoselect(
+                        PossibleValues => $NextStates,
+                    ) || '';
+                    if ( $GetParam{StateID} ) {
+                        $NewChangedElements{NextStateID} = 1;
+                        $ChangedStdFields{NextStateID}   = 1;
+                    }
+                }
+
+                if ( !%NewChangedElements ) {
+                    $Convergence{StdFields} = 1;
+                }
+                else {
+                    %ChangedElements = %NewChangedElements;
+                }
+
+                %ChangedElementsDFStart = (
+                    %ChangedElementsDFStart,
+                    %NewChangedElements,
+                );
+
+                if ( $LoopProtection-- < 1 ) {
+                    $Kernel::OM->Get('Kernel::System::Log')->Log(
+                        Priority => 'error',
+                        Message  => "Ran into unresolvable loop!",
+                    );
+                    return;
+                }
 
-                # convert Filer key => key back to key => value using map
-                %{$PossibleValues} = map { $_ => $PossibleValues->{$_} } keys %Filter;
             }
 
-            my $DataValues = $DynamicFieldBackendObject->BuildSelectionDataGet(
-                DynamicFieldConfig => $DynamicFieldConfig,
-                PossibleValues     => $PossibleValues,
-                Value              => $DynamicFieldValues{ $DynamicFieldConfig->{Name} },
-            ) || $PossibleValues;
+            %ChangedElements        = %ChangedElementsDFStart;
+            %ChangedElementsDFStart = ();
+
+            # check dynamic fields
+            my %CurFieldStates;
+            if (%ChangedElements) {
+
+                # get values and visibility of dynamic fields
+                %CurFieldStates = $FieldRestrictionsObject->GetFieldStates(
+                    TicketObject              => $TicketObject,
+                    DynamicFields             => $Self->{DynamicField},
+                    DynamicFieldBackendObject => $DynamicFieldBackendObject,
+                    ChangedElements           => \%ChangedElements,            # optional to reduce ACL evaluation
+                    Action                    => $Self->{Action},
+                    UserID                    => $Self->{UserID},
+                    FormID                    => $Self->{FormID},
+                    CustomerUser              => $Selected || '',
+                    GetParam                  => {
+                        %GetParam,
+                        NextStateID => $GetParam{StateID},
```

Comment:
When aligning the things, NextStateID should already be used in the code, so no need for this anymore, too.